src/java/org/apache/cassandra/cql3/restrictions/LikePattern.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.cql3.restrictions;
+
+import java.nio.ByteBuffer;
+
+import org.apache.cassandra.cql3.Operator;
+import org.apache.cassandra.utils.ByteBufferUtil;
+
+import static org.apache.cassandra.cql3.statements.RequestValidations.invalidRequest;
+
+public final class LikePattern
```

Comment:
At the moment, I guess the SASI tests make sure this still works, but at some point in the future, when there's support for LIKE via filtering, SAI, etc. there will be others.

Reply:
I strongly feel that we should revisit the LIKE implementation. The `isSatisfiedBy` methods of the LIKE operators are wrong as they do not account for the fact that LIKE is case insensitive. We just never noticed because the code is never used.
Also the approach of creating multiple operators does not bring much considering how quick is the computation.

Reply:
Case-sensitivity is something the index implementations certainly can take care of, but if we support it in filtering space, there would have to be something like a argument to the LIKE operator that specifies the behavior.

Reply:
I thought that the default for SASI was to be case insensitive with LIKE. If it is the case we should have a similar behaviour with filtering. One major problem with CQL is that we have all those difference in behaviour. I would love to have something harmonized in term of behaviour.

Reply:
The default is case-sensitive, but both SAI and SASI can enable case-insensitivity via filtering/analysis.